File: leetcode/0141-linked-list-cycle.cc
// https://leetcode.com/problems/linked-list-cycle/
// Easy

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* slow = head;
        ListNode* fast = head->next;
        while (slow && fast) {
            if (slow == fast) return true;
            if (fast->next && fast->next->next) {
                fast = fast->next->next;
            } else {
                return false;
            }
            slow = slow->next;
        }
        return false;
    }
};

File: leetcode/python/0141-linked-list-cycle.py
# https://leetcode.com/problems/linked-list-cycle/
# Medium

# Have one fast and one slow node
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        fast, slow = head.next if head else None, head
        while fast:
            if slow is fast:
                return True
            fast = fast.next.next if fast.next else None
            slow = slow.next
        return False